File: 2022/day3.clj
(require '[clojure.string :as str])
(require '[clojure.set])
(require '[clojure.zip :as zip])

(def elves (str/split-lines (slurp "input/3")))
(defn priority [char]
    (let [
        c (int char)
        a (int \a)
        z (int \z)
        A (int \A)
    ]
    (if (<= a c z)
        (+ (- c a) 1)
        (+ (- c A) 27))))

; Part 1
(defn two-knapsacks [line] (
    let [
        half (/ (count line) 2)
        [knapsack-1 knapsack-2] (map set (split-at half line))
    ]
    (priority (first (clojure.set/intersection knapsack-1 knapsack-2)
))))
(println (reduce + (map two-knapsacks elves)))

; Part 2
(defn chunker [sequence num]
    (if (<= (count sequence) num)
    [sequence]
    (let [[head tail] (split-at num sequence)] (lazy-seq (cons head (chunker tail num))))))
(defn three-elves [lines]
    (priority (first (apply clojure.set/intersection (map set lines)))))
(println (reduce + (map three-elves (chunker elves 3))))

File: 2022/day3.scala
import scala.io.Source

extension (s: String) def splitHalf: (String, String) = (s.substring(0, s.length()/2), s.substring(s.length()/2))

def priorityScore(c: Char): Int = c match
	case c if 'a' to 'z' contains c => (c - 'a' + 1)
	case c if 'A' to 'Z' contains c => (c - 'A' + 27)
	case _ => 0


@main def main() =
	val inputLines = Source.fromFile("input/3").getLines.toArray  // Can't leave it lazy as Part 1 will consume it

	// Part 1 - find items common between first half and second half of each line
	println(inputLines.map(line =>
		val (knapsack1, knapsack2) = line.splitHalf
		priorityScore((knapsack1.toSet & knapsack2.toSet).head)
	).sum)

	// Part 2 - find items common between lines in groups of 3 lines
	println(inputLines.grouped(3).map(group =>
		priorityScore(group.map(_.toSet).reduce(_&_).head)
	).sum)
